compiler/rustc_type_ir/src/structural_impls.rs

```diff
@@ -173,3 +192,15 @@ impl<I: Interner, T: TypeVisitable<I>, Ix: Idx> TypeVisitable<I> for IndexVec<Ix
         self.iter().try_for_each(|t| t.visit_with(visitor))
     }
 }
+
+impl<'a, I: Interner, T: Internable<'a, I> + TypeFoldable<I>> TypeFoldable<I> for Interned<'a, T> {
+    fn try_fold_with<F: FallibleTypeFolder<I>>(self, folder: &mut F) -> Result<Self, F::Error> {
+        (*self).clone().try_fold_with(folder).map(|v| v.intern(folder.interner()))
```

Comment:
i feel like `Interned` shouldn't implement `TypeFoldable` in a generic way.

Types which are interned tend to be hot and at least should have the "if fold noop => do not try to intern again" check. Which types need this impl? This might be perf relevant

Reply:
> Which types need this impl? This might be perf relevant

This is (currently) used only by `ExternalConstraints`, the incumbent folding of which interns a new instance already.

Reply:
Actually maybe that's not right.  I think 23baca779750d6c07f6c8ebb15bdd36076b0faaa expanded it to other types.

Reply:
I've not changed this yet since it's the same as the status quo, so unlikely responsible for perf regression.  It's not `#[inline]` however, so that could be relevant?  Leaving conversation unresolved for now.